# 📊 MLflow Experiment Tracking — Used Car Price Prediction

This notebook demonstrates **end-to-end experiment tracking and model management** using [MLflow](https://mlflow.org/) for our Used Car Price Prediction & Valuation System.

---

## What is MLflow?

MLflow is an **open-source platform** for managing the complete machine learning lifecycle. It provides four core components:

| Component | Purpose |
|-----------|--------|
| **MLflow Tracking** | Log parameters, metrics, and artifacts for each experiment run |
| **MLflow Projects** | Package ML code in a reproducible, reusable format |
| **MLflow Models** | Deploy models in diverse serving environments |
| **MLflow Model Registry** | Centralized model store with versioning, staging, and production lifecycle |

## Why is Experiment Tracking Important?

In real-world ML projects, data scientists run **hundreds of experiments** — different algorithms, hyperparameters, feature sets, and preprocessing strategies. Without systematic tracking:

- ❌ You forget which configuration produced the best model
- ❌ Results are not reproducible
- ❌ Team collaboration is chaotic
- ❌ Auditing and compliance are impossible

With MLflow:

- ✅ Every run is logged with parameters, metrics, and artifacts
- ✅ Results are queryable, comparable, and sortable
- ✅ Models are versioned and traceable from experiment to production
- ✅ The entire ML lifecycle is governed and auditable

## How Model Versioning Works

MLflow's **Model Registry** enables:

1. **Registration** — Save a trained model with a unique name
2. **Versioning** — Each retrained model gets a new version (v1, v2, v3...)
3. **Stage Transitions** — Move models through stages: `None → Staging → Production → Archived`
4. **Annotations** — Add descriptions and tags to each version for documentation

This ensures that the model serving in production is always **traceable** back to the exact experiment, dataset, and code that created it.

---

### Notebook Roadmap

| Section | Purpose |
|---------|--------|
| 1 | Install & Configure MLflow |
| 2 | Load Data & Prepare Train/Test Split |
| 3 | Create MLflow Experiment |
| 4 | Train & Log — Linear Regression |
| 5 | Train & Log — Random Forest Regressor |
| 6 | Train & Log — Tuned XGBoost Regressor (Champion) |
| 7 | Experiment Comparison — All Runs |
| 8 | Register Champion Model in MLflow Model Registry |
| 9 | Why XGBoost Was Selected as the Champion Model |
| 10 | Launching the MLflow UI |
| 11 | Final Summary |

---
## 1 · Install & Configure MLflow

In [1]:
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')  # Non-interactive backend for saving plots
import matplotlib.pyplot as plt
import seaborn as sns
import shap
import mlflow
import mlflow.sklearn
import mlflow.xgboost
import warnings
import time
import os
import tempfile

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from xgboost import XGBRegressor

warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

sns.set_style('whitegrid')
plt.rcParams.update({
    'figure.figsize': (14, 7),
    'font.size': 12,
    'axes.titlesize': 14,
    'axes.labelsize': 12,
    'figure.dpi': 120,
})

RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)

print(f'MLflow version : {mlflow.__version__}')
print(f'SHAP version   : {shap.__version__}')
print('All libraries imported successfully.')

c:\Users\admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


MLflow version : 3.13.0
SHAP version   : 0.52.0
All libraries imported successfully.


### MLflow Configuration

We configure MLflow to use a **local SQLite database** as the tracking backend. This is the recommended approach for local development — the older filesystem backend (`./mlruns`) is now deprecated in recent MLflow versions. SQLite also enables **full Model Registry support**, which the file-based backend lacked.

In production, you would point this to a remote tracking server (e.g., Databricks, AWS, or a self-hosted MLflow server with PostgreSQL).

```
Project Directory
├── mlflow.db            ← SQLite database (tracking store)
├── mlflow-artifacts/    ← Default artifact store
│   ├── <experiment_id>/
│   │   ├── <run_id>/
│   │   │   └── artifacts/
│   │   │       ├── plots/
│   │   │       │   ├── feature_importance.png
│   │   │       │   └── shap_summary.png
│   │   │       └── model/
│   │   └── ...
│   └── ...
└── ...
```

In [2]:
# ── Configure MLflow tracking URI (local SQLite database) ──
# The filesystem backend ('mlruns') is deprecated in recent MLflow versions.
# SQLite is the recommended lightweight local backend and supports the full Model Registry.
MLFLOW_TRACKING_URI = 'sqlite:///mlflow.db'
mlflow.set_tracking_uri(MLFLOW_TRACKING_URI)

print(f'MLflow tracking URI : {mlflow.get_tracking_uri()}')
print('✅ MLflow configured successfully with SQLite backend.')

MLflow tracking URI : sqlite:///mlflow.db
✅ MLflow configured successfully with SQLite backend.


---
## 2 · Load Data & Prepare Train/Test Split

We use the same **feature-engineered dataset** and **identical train-test split** (`test_size=0.20`, `random_state=42`) as the Baseline_Models and Hyperparameter_Tuning notebooks for fair, reproducible comparisons.

In [3]:
# ── Load dataset ──
df = pd.read_csv('feature_engineered_car_data.csv')
print(f'Dataset shape : {df.shape}')
print(f'Null values   : {df.isnull().sum().sum()}')

# ── Define target & features ──
TARGET = 'listed_price'
drop_cols = [TARGET, 'discountValue']

X = df.drop(columns=drop_cols)
y = df[TARGET]

# ── Train-test split ──
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.20, random_state=RANDOM_STATE
)

print(f'\nFeatures      : {X.shape[1]}')
print(f'Training set  : {X_train.shape[0]:,} samples')
print(f'Test set      : {X_test.shape[0]:,} samples')

# Store feature names for later use
feature_names = list(X.columns)

Dataset shape : (37750, 83)
Null values   : 0

Features      : 81
Training set  : 30,200 samples
Test set      : 7,550 samples


---
## 3 · Create MLflow Experiment

An **MLflow Experiment** groups related runs together. All our model training runs will be logged under the experiment named `Used_Car_Price_Prediction`.

In [4]:
# ── Create / Set the MLflow Experiment ──
EXPERIMENT_NAME = 'Used_Car_Price_Prediction'

experiment = mlflow.set_experiment(EXPERIMENT_NAME)

print(f'Experiment Name : {experiment.name}')
print(f'Experiment ID   : {experiment.experiment_id}')
print(f'Artifact Location : {experiment.artifact_location}')
print(f'Lifecycle Stage   : {experiment.lifecycle_stage}')
print('\n✅ MLflow experiment created/set successfully.')

Experiment Name : Used_Car_Price_Prediction
Experiment ID   : 1
Artifact Location : file:d:/MLPs/Used Car Price Prediction & Valuation System/mlruns/1
Lifecycle Stage   : active

✅ MLflow experiment created/set successfully.


---
## Helper Functions

We define reusable helper functions for:
- Computing and logging metrics
- Generating and logging feature importance plots
- Generating and logging SHAP summary plots

In [5]:
def compute_metrics(y_true, y_pred):
    """Compute MAE, RMSE, and R² score."""
    mae  = mean_absolute_error(y_true, y_pred)
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    r2   = r2_score(y_true, y_pred)
    return {'MAE': mae, 'RMSE': rmse, 'R2': r2}


def log_metrics_to_mlflow(metrics_dict, prefix='test'):
    """Log a dictionary of metrics to the active MLflow run."""
    for name, value in metrics_dict.items():
        mlflow.log_metric(f'{prefix}_{name}', value)


def create_and_log_feature_importance(model, feature_names, model_name, top_n=20):
    """Generate a feature importance bar plot and log it as an MLflow artifact."""
    try:
        importances = model.feature_importances_
    except AttributeError:
        # LinearRegression uses coef_ instead of feature_importances_
        importances = np.abs(model.coef_)

    imp_df = pd.DataFrame({
        'Feature': feature_names,
        'Importance': importances
    }).sort_values('Importance', ascending=False).head(top_n)

    fig, ax = plt.subplots(figsize=(10, 8))
    ax.barh(imp_df['Feature'][::-1], imp_df['Importance'][::-1],
            color='#3498db', edgecolor='#1a1a2e', linewidth=0.5)
    ax.set_xlabel('Importance', fontsize=12)
    ax.set_title(f'Feature Importance — {model_name} (Top {top_n})',
                 fontsize=13, fontweight='bold')
    for spine in ax.spines.values():
        spine.set_visible(False)
    ax.tick_params(left=False)
    plt.tight_layout()

    # Save and log
    filepath = os.path.join(tempfile.gettempdir(), f'feature_importance_{model_name.replace(" ", "_")}.png')
    fig.savefig(filepath, dpi=150, bbox_inches='tight')
    plt.close(fig)
    mlflow.log_artifact(filepath, artifact_path='plots')
    print(f'  📊 Feature importance plot logged.')
    return filepath


def create_and_log_shap_summary(model, X_data, model_name, max_display=20):
    """Generate a SHAP summary plot and log it as an MLflow artifact."""
    try:
        explainer = shap.TreeExplainer(model)
        shap_values = explainer.shap_values(X_data)
    except Exception:
        # Fallback for non-tree models (e.g., LinearRegression)
        try:
            # Use a small background sample for KernelExplainer
            background = shap.sample(X_data, min(100, len(X_data)))
            explainer = shap.LinearExplainer(model, background)
            shap_values = explainer.shap_values(X_data.iloc[:200])  # Limit for speed
            X_data = X_data.iloc[:200]
        except Exception as e:
            print(f'  ⚠️ SHAP computation failed for {model_name}: {e}')
            return None

    fig, ax = plt.subplots(figsize=(12, 8))
    shap.summary_plot(
        shap_values,
        X_data,
        max_display=max_display,
        show=False,
        plot_size=None
    )
    plt.title(f'SHAP Summary — {model_name}',
              fontsize=13, fontweight='bold', pad=15)
    plt.tight_layout()

    filepath = os.path.join(tempfile.gettempdir(), f'shap_summary_{model_name.replace(" ", "_")}.png')
    plt.savefig(filepath, dpi=150, bbox_inches='tight')
    plt.close('all')
    mlflow.log_artifact(filepath, artifact_path='plots')
    print(f'  🔍 SHAP summary plot logged.')
    return filepath


def print_run_summary(run_name, params, train_metrics, test_metrics, training_time):
    """Print a formatted summary of the run."""
    print(f'\n{"=" * 65}')
    print(f'  {run_name} — Run Summary')
    print(f'{"=" * 65}')
    print(f'  Training Time : {training_time:.2f} seconds')
    print(f'  {"":15s} {"Train":>15s} {"Test":>15s}')
    print(f'  {"-" * 50}')
    print(f'  {"MAE":15s} {train_metrics["MAE"]:>15,.2f} {test_metrics["MAE"]:>15,.2f}')
    print(f'  {"RMSE":15s} {train_metrics["RMSE"]:>15,.2f} {test_metrics["RMSE"]:>15,.2f}')
    print(f'  {"R²":15s} {train_metrics["R2"]:>15.4f} {test_metrics["R2"]:>15.4f}')
    print(f'{"=" * 65}')


print('✅ Helper functions defined.')

✅ Helper functions defined.


---
## 4 · Train & Log — Linear Regression

Linear Regression serves as our **simplest baseline**. It assumes a linear relationship between features and the target price. While it won't capture the complex non-linear interactions in used car pricing, it provides a useful lower bound for comparison.

**What we log to MLflow:**
- Model parameters
- Train & Test metrics (MAE, RMSE, R²)
- Training time
- Feature importance plot (using absolute coefficient values)
- SHAP summary plot
- The trained model artifact

In [6]:
# ── Run 1: Linear Regression ──
RUN_NAME_LR = 'Linear Regression'

with mlflow.start_run(run_name=RUN_NAME_LR) as run_lr:
    print(f'🚀 Starting run: {RUN_NAME_LR}')
    print(f'   Run ID: {run_lr.info.run_id}')

    # ── Define & log parameters ──
    lr_params = {
        'model_type': 'LinearRegression',
        'fit_intercept': True,
        'n_features': X_train.shape[1],
        'train_samples': X_train.shape[0],
        'test_samples': X_test.shape[0],
    }
    mlflow.log_params(lr_params)

    # ── Train ──
    start_time = time.time()
    lr_model = LinearRegression(fit_intercept=True)
    lr_model.fit(X_train, y_train)
    lr_training_time = time.time() - start_time

    # ── Predict ──
    y_train_pred_lr = lr_model.predict(X_train)
    y_test_pred_lr  = lr_model.predict(X_test)

    # ── Compute & log metrics ──
    lr_train_metrics = compute_metrics(y_train, y_train_pred_lr)
    lr_test_metrics  = compute_metrics(y_test, y_test_pred_lr)

    log_metrics_to_mlflow(lr_train_metrics, prefix='train')
    log_metrics_to_mlflow(lr_test_metrics, prefix='test')
    mlflow.log_metric('training_time_seconds', lr_training_time)

    # ── Log feature importance plot ──
    create_and_log_feature_importance(lr_model, feature_names, RUN_NAME_LR)

    # ── Log SHAP summary plot ──
    create_and_log_shap_summary(lr_model, X_test, RUN_NAME_LR)

    # ── Log model artifact ──
    mlflow.sklearn.log_model(lr_model, artifact_path='model')
    print(f'  💾 Model artifact logged.')

    # ── Tags ──
    mlflow.set_tag('model_family', 'Linear')
    mlflow.set_tag('stage', 'baseline')
    mlflow.set_tag('dataset', 'feature_engineered_car_data.csv')

    # ── Print summary ──
    print_run_summary(RUN_NAME_LR, lr_params, lr_train_metrics, lr_test_metrics, lr_training_time)

# Store run ID for comparison
lr_run_id = run_lr.info.run_id
print(f'\n✅ Run completed. Run ID: {lr_run_id}')

2026/06/21 16:44:13 WARNING mlflow.utils.git_utils: Failed to import Git (the Git executable is probably not on your PATH), so Git SHA is not available. Error: Failed to initialize: Bad git executable.
The git executable must be specified in one of the following ways:
    - be included in your $PATH
    - be set via $GIT_PYTHON_GIT_EXECUTABLE
    - explicitly set via git.refresh(<full-path-to-git-executable>)

All git commands will error until this is rectified.

This initial message can be silenced or aggravated in the future by setting the
$GIT_PYTHON_REFRESH environment variable. Use one of the following values:
    - quiet|q|silence|s|silent|none|n|0: for no message or exception
    - warn|w|warning|log|l|1: for a warning message (logging level CRITICAL, displayed by default)
    - error|e|exception|raise|r|2: for a raised exception

Example:
    export GIT_PYTHON_REFRESH=quiet



🚀 Starting run: Linear Regression
   Run ID: 45ae6670b4c64281a0b964c71fdd23d0
  📊 Feature importance plot logged.


2026/06/21 16:44:16 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/21 16:44:17 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html


  🔍 SHAP summary plot logged.
  💾 Model artifact logged.

  Linear Regression — Run Summary
  Training Time : 0.23 seconds
                            Train            Test
  --------------------------------------------------
  MAE                  116,208.46      118,221.01
  RMSE                 157,338.87      160,143.83
  R²                       0.8798          0.8760

✅ Run completed. Run ID: 45ae6670b4c64281a0b964c71fdd23d0


### 📝 Observations — Linear Regression

- Linear Regression provides a **fast baseline** with near-instantaneous training time.
- The R² score is decent but significantly lower than tree-based models — this is expected because used car pricing involves **highly non-linear relationships** (e.g., depreciation curves, brand interactions).
- The large gap between Train and Test MAE (if present) may indicate that some linear assumptions are violated.
- This baseline helps us quantify the **improvement** that more complex models bring.

---
## 6 · Train & Log — Tuned XGBoost Regressor (Champion Model)

This is our **champion model** — the Tuned XGBoost Regressor with hyperparameters optimized via a two-phase strategy:

| Phase | Method | Purpose |
|-------|--------|---------|
| 1 | RandomizedSearchCV (50 iter × 3-fold) | Broad exploration of 7 hyperparameters |
| 2 | GridSearchCV (30 combos × 3-fold) | Fine-tune top 3 most impactful parameters |

**Best hyperparameters discovered:**

| Parameter | Value | Tuning Status |
|-----------|-------|---------------|
| `n_estimators` | 500 | Tuned (Phase 2) |
| `max_depth` | 8 | Fixed (Phase 1) |
| `learning_rate` | 0.03 | Tuned (Phase 2) |
| `subsample` | 0.7 | Fixed (Phase 1) |
| `colsample_bytree` | 0.9 | Fixed (Phase 1) |
| `min_child_weight` | 1 | Tuned (Phase 2) |
| `gamma` | 1.0 | Fixed (Phase 1) |

In [7]:
# ── Run 3: Tuned XGBoost Regressor (Champion) ──
RUN_NAME_XGB = 'Tuned XGBoost Regressor'

with mlflow.start_run(run_name=RUN_NAME_XGB) as run_xgb:
    print(f'🚀 Starting run: {RUN_NAME_XGB}')
    print(f'   Run ID: {run_xgb.info.run_id}')

    # ── Define & log parameters ──
    xgb_params = {
        'model_type': 'XGBRegressor',
        'n_estimators': 500,
        'max_depth': 8,
        'learning_rate': 0.03,
        'subsample': 0.7,
        'colsample_bytree': 0.9,
        'min_child_weight': 1,
        'gamma': 1.0,
        'reg_alpha': 0.1,
        'reg_lambda': 1.0,
        'random_state': RANDOM_STATE,
        'tuning_strategy': 'RandomizedSearchCV + GridSearchCV',
        'n_features': X_train.shape[1],
        'train_samples': X_train.shape[0],
        'test_samples': X_test.shape[0],
    }
    mlflow.log_params(xgb_params)

    # ── Train ──
    start_time = time.time()
    xgb_model = XGBRegressor(
        n_estimators=500,
        max_depth=8,
        learning_rate=0.03,
        subsample=0.7,
        colsample_bytree=0.9,
        min_child_weight=1,
        gamma=1.0,
        reg_alpha=0.1,
        reg_lambda=1.0,
        n_jobs=-1,
        random_state=RANDOM_STATE,
        verbosity=0
    )
    xgb_model.fit(X_train, y_train)
    xgb_training_time = time.time() - start_time

    # ── Predict ──
    y_train_pred_xgb = xgb_model.predict(X_train)
    y_test_pred_xgb  = xgb_model.predict(X_test)

    # ── Compute & log metrics ──
    xgb_train_metrics = compute_metrics(y_train, y_train_pred_xgb)
    xgb_test_metrics  = compute_metrics(y_test, y_test_pred_xgb)

    log_metrics_to_mlflow(xgb_train_metrics, prefix='train')
    log_metrics_to_mlflow(xgb_test_metrics, prefix='test')
    mlflow.log_metric('training_time_seconds', xgb_training_time)

    # ── Log feature importance plot ──
    create_and_log_feature_importance(xgb_model, feature_names, RUN_NAME_XGB)

    # ── Log SHAP summary plot ──
    create_and_log_shap_summary(xgb_model, X_test, RUN_NAME_XGB)

    # ── Log model artifact (using mlflow.xgboost for native XGBoost support) ──
    mlflow.xgboost.log_model(xgb_model, artifact_path='model')
    print(f'  💾 Model artifact logged (XGBoost native format).')

    # ── Tags ──
    mlflow.set_tag('model_family', 'Ensemble - Boosting')
    mlflow.set_tag('stage', 'champion')
    mlflow.set_tag('dataset', 'feature_engineered_car_data.csv')
    mlflow.set_tag('champion_model', 'True')
    mlflow.set_tag('tuning_method', 'Phase1-RandomizedSearchCV_Phase2-GridSearchCV')

    # ── Print summary ──
    print_run_summary(RUN_NAME_XGB, xgb_params, xgb_train_metrics, xgb_test_metrics, xgb_training_time)

# Store run ID for comparison
xgb_run_id = run_xgb.info.run_id
print(f'\n✅ Run completed. Run ID: {xgb_run_id}')

🚀 Starting run: Tuned XGBoost Regressor
   Run ID: 9a465c22ca5c4a80b70bd5db33f9c40e
  📊 Feature importance plot logged.


2026/06/21 16:47:13 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.


  🔍 SHAP summary plot logged.
  💾 Model artifact logged (XGBoost native format).

  Tuned XGBoost Regressor — Run Summary
  Training Time : 12.94 seconds
                            Train            Test
  --------------------------------------------------
  MAE                   35,216.84       54,060.33
  RMSE                  49,178.49       83,343.83
  R²                       0.9883          0.9664

✅ Run completed. Run ID: 9a465c22ca5c4a80b70bd5db33f9c40e


### 📝 Observations — Tuned XGBoost Regressor

- The Tuned XGBoost achieves the **best test R²** (~0.9664) among all models, confirming it as the champion.
- **Reduced overfitting** compared to the baseline XGBoost (from the Hyperparameter_Tuning notebook): the gap between train and test metrics is smaller thanks to regularization (`gamma=1.0`, `reg_alpha=0.1`, `reg_lambda=1.0`) and subsampling (`subsample=0.7`).
- **Lower learning rate** (`0.03` vs `0.1` baseline) with more trees (`500` vs `300`) allows the model to learn more gradually and generalize better.
- Training time is moderate — acceptable for batch retraining in production.

---
## 7 · Experiment Comparison — All Runs

Now we retrieve all runs from our MLflow experiment and create a **comparison table** to evaluate which model performs best.

In [8]:
# ── Retrieve all runs from the experiment ──
runs_df = mlflow.search_runs(
    experiment_ids=[experiment.experiment_id],
    order_by=['metrics.test_R2 DESC']
)

print(f'Total runs in experiment "{EXPERIMENT_NAME}": {len(runs_df)}')
print(f'\nAvailable columns: {list(runs_df.columns)[:10]}...')

Total runs in experiment "Used_Car_Price_Prediction": 6

Available columns: ['run_id', 'experiment_id', 'status', 'artifact_uri', 'start_time', 'end_time', 'metrics.train_RMSE', 'metrics.training_time_seconds', 'metrics.train_R2', 'metrics.test_RMSE']...


In [9]:
# ── Build a clean comparison table ──
comparison_cols = {
    'tags.mlflow.runName': 'Model',
    'params.model_type': 'Algorithm',
    'metrics.train_MAE': 'Train MAE',
    'metrics.train_RMSE': 'Train RMSE',
    'metrics.train_R2': 'Train R²',
    'metrics.test_MAE': 'Test MAE',
    'metrics.test_RMSE': 'Test RMSE',
    'metrics.test_R2': 'Test R²',
    'metrics.training_time_seconds': 'Training Time (s)',
    'tags.stage': 'Stage',
}

# Filter to only columns that exist
available_cols = {k: v for k, v in comparison_cols.items() if k in runs_df.columns}
comparison = runs_df[list(available_cols.keys())].rename(columns=available_cols)

# Sort by Test R² (descending)
if 'Test R²' in comparison.columns:
    comparison = comparison.sort_values('Test R²', ascending=False).reset_index(drop=True)

print('=' * 100)
print('  EXPERIMENT COMPARISON — All Runs (Sorted by Test R²)')
print('=' * 100)
print(comparison.to_string(index=False))
print('=' * 100)

  EXPERIMENT COMPARISON — All Runs (Sorted by Test R²)
                  Model             Algorithm     Train MAE    Train RMSE  Train R²      Test MAE     Test RMSE  Test R²  Training Time (s)    Stage
Tuned XGBoost Regressor          XGBRegressor  35216.844422  49178.493685  0.988257  54060.333921  83343.831114 0.966417          12.937051 champion
Random Forest Regressor RandomForestRegressor  51346.262801  73686.399672  0.973637  66275.947026  99105.422603 0.952513           9.985321     None
Random Forest Regressor RandomForestRegressor  51518.429426  74042.955430  0.973381  66416.599004  99191.992322 0.952430           3.877322     None
      Linear Regression      LinearRegression 116208.455135 157338.867706  0.879803 118221.010036 160143.825505 0.876006           0.233685 baseline
      Linear Regression      LinearRegression 116208.455135 157338.867706  0.879803 118221.010036 160143.825505 0.876006           0.374706 baseline
      Linear Regression      LinearRegression 11620

In [10]:
# ── Visual Comparison: Test Metrics ──
if 'Model' in comparison.columns and 'Test R²' in comparison.columns:
    fig, axes = plt.subplots(1, 3, figsize=(18, 6))

    metrics_to_plot = [
        ('Test MAE', 'Mean Absolute Error (MAE)', '#EF5350'),
        ('Test RMSE', 'Root Mean Squared Error (RMSE)', '#FFA726'),
        ('Test R²', 'R² Score', '#42A5F5'),
    ]

    for idx, (metric_col, title, color) in enumerate(metrics_to_plot):
        if metric_col in comparison.columns:
            models = comparison['Model'].tolist()
            values = comparison[metric_col].tolist()

            bars = axes[idx].bar(models, values, color=color,
                                edgecolor='#1a1a2e', linewidth=0.5, width=0.5)
            axes[idx].set_title(title, fontweight='bold', fontsize=12)
            axes[idx].set_ylabel(metric_col, fontsize=11)

            for bar, v in zip(bars, values):
                if metric_col == 'Test R²':
                    axes[idx].text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.005,
                                  f'{v:.4f}', ha='center', va='bottom', fontweight='bold', fontsize=10)
                else:
                    axes[idx].text(bar.get_x() + bar.get_width()/2, bar.get_height() + 500,
                                  f'{v:,.0f}', ha='center', va='bottom', fontweight='bold', fontsize=10)

            axes[idx].tick_params(axis='x', rotation=15)

    plt.suptitle('Model Comparison — Test Set Performance (MLflow Tracked)',
                 fontsize=15, fontweight='bold', y=1.03)
    plt.tight_layout()
    plt.show()
else:
    print('⚠️ Comparison columns not available for plotting.')

### 📝 Observations — Experiment Comparison

The comparison table and charts clearly show the **performance progression**:

| Model | Strengths | Weaknesses |
|-------|-----------|------------|
| **Linear Regression** | Fastest training, simplest to interpret | Lowest accuracy — cannot capture non-linear pricing patterns |
| **Random Forest** | Strong non-linear capture, robust to outliers | Higher training time, potential overfitting, no gradient optimization |
| **Tuned XGBoost** ✅ | Best accuracy (R², MAE, RMSE), well-regularized, tuned hyperparameters | Slightly more complex, moderate training time |

The **Tuned XGBoost Regressor** is the clear winner and is selected as the **champion model** for production deployment.

---
## 8 · Register Champion Model in MLflow Model Registry

The **MLflow Model Registry** provides a centralized model store for managing the lifecycle of ML models. We register our champion (Tuned XGBoost) model here.

### Model Registry Workflow:
```
Experiment Run → Register Model → Version 1 → Stage: None
                                                  ↓
                                           Stage: Staging
                                                  ↓
                                           Stage: Production ✅
                                                  ↓
                                           Stage: Archived
```

In [11]:
# ── Register the Champion Model ──
MODEL_REGISTRY_NAME = 'Used_Car_Price_XGBoost_Champion'

model_uri = f'runs:/{xgb_run_id}/model'

print(f'Registering model from run: {xgb_run_id}')
print(f'Model URI: {model_uri}')
print(f'Registry Name: {MODEL_REGISTRY_NAME}')

try:
    # Register the model
    registered_model = mlflow.register_model(
        model_uri=model_uri,
        name=MODEL_REGISTRY_NAME
    )

    print(f'\n✅ Model registered successfully!')
    print(f'   Name    : {registered_model.name}')
    print(f'   Version : {registered_model.version}')
    print(f'   Source  : {registered_model.source}')
    print(f'   Status  : {registered_model.status}')

except Exception as e:
    print(f'\n⚠️ Model registration note: {e}')
    print('   In a production setup (e.g., Databricks, MLflow Server with DB backend), registration works seamlessly.')

Registering model from run: 9a465c22ca5c4a80b70bd5db33f9c40e
Model URI: runs:/9a465c22ca5c4a80b70bd5db33f9c40e/model
Registry Name: Used_Car_Price_XGBoost_Champion


Successfully registered model 'Used_Car_Price_XGBoost_Champion'.
2026/06/21 16:47:25 WARNING mlflow.tracking._model_registry.fluent: Run with id 9a465c22ca5c4a80b70bd5db33f9c40e has no artifacts at artifact path 'model', registering model based on models:/m-9a401bf4fb7f4257810579f48c8c4736 instead



✅ Model registered successfully!
   Name    : Used_Car_Price_XGBoost_Champion
   Version : 1
   Source  : models:/m-9a401bf4fb7f4257810579f48c8c4736
   Status  : READY


Created version '1' of model 'Used_Car_Price_XGBoost_Champion'.


In [12]:
# ── Describe the registered model ──
try:
    from mlflow.tracking import MlflowClient

    client = MlflowClient()

    # Update model description
    client.update_registered_model(
        name=MODEL_REGISTRY_NAME,
        description=(
            'Champion model for Used Car Price Prediction. '
            'Tuned XGBoost Regressor with hyperparameters optimized via '
            'RandomizedSearchCV (Phase 1) + GridSearchCV (Phase 2). '
            f'Test R² = {xgb_test_metrics["R2"]:.4f}, '
            f'Test MAE = {xgb_test_metrics["MAE"]:,.2f}, '
            f'Test RMSE = {xgb_test_metrics["RMSE"]:,.2f}. '
            'Trained on feature_engineered_car_data.csv (37,750 samples, 81 features).'
        )
    )

    # Update version description
    client.update_model_version(
        name=MODEL_REGISTRY_NAME,
        version=registered_model.version,
        description=(
            f'Version {registered_model.version}: Initial champion model. '
            f'Tuned XGBoost (n_estimators=500, lr=0.03, max_depth=8). '
            f'Test R²={xgb_test_metrics["R2"]:.4f}.'
        )
    )

    print('✅ Model and version descriptions updated.')

    # ── List all registered models ──
    print('\n── Registered Models ──')
    for rm in client.search_registered_models():
        print(f'  Name: {rm.name}')
        for mv in rm.latest_versions:
            print(f'    Version: {mv.version} | Stage: {mv.current_stage} | Status: {mv.status}')

except Exception as e:
    print(f'⚠️ Registry operations note: {e}')
    print('   Full registry features require a database-backed MLflow server.')

✅ Model and version descriptions updated.

── Registered Models ──
  Name: Used_Car_Price_XGBoost_Champion
    Version: 1 | Stage: None | Status: READY


### 📝 Observations — Model Registration

The champion model (Tuned XGBoost Regressor) is now registered in the MLflow Model Registry with:

- **A unique name**: `Used_Car_Price_XGBoost_Champion`
- **Version 1**: The initial production-ready model
- **Full traceability**: Links back to the exact experiment run, parameters, metrics, and artifacts
- **Descriptions**: Both the model and version are annotated with performance metrics and training details

In a production environment, you would:
1. Transition this version to `Staging` for integration testing
2. Run validation tests against a holdout dataset
3. Promote to `Production` when validated
4. Archive older versions as new ones are trained

---
## 9 · Why XGBoost Was Selected as the Champion Model

The selection of the **Tuned XGBoost Regressor** as the champion model is based on a systematic evaluation across multiple dimensions:

### 1. Superior Predictive Accuracy

| Metric | Linear Regression | Random Forest | Tuned XGBoost ✅ |
|--------|:-:|:-:|:-:|
| Test R² | Lowest | Mid | **Highest (~0.9664)** |
| Test MAE | Highest | Mid | **Lowest (~₹54,060)** |
| Test RMSE | Highest | Mid | **Lowest (~₹83,344)** |

XGBoost delivers the **best generalization performance** on unseen test data across all three key metrics.

### 2. Robust Generalization (Low Overfitting)

Despite being the most complex model, the Tuned XGBoost has a **controlled train-test gap** thanks to:
- **Gamma = 1.0** — Minimum loss reduction for further partition (tree complexity penalty)
- **Subsample = 0.7** — Only 70% of training data per tree (stochastic regularization)
- **L1/L2 regularization** (`reg_alpha=0.1`, `reg_lambda=1.0`)
- **Lower learning rate** (`0.03`) with more trees (`500`) — gradual, stable learning

### 3. Systematic Hyperparameter Tuning

Unlike the other models (used with defaults or basic configs), XGBoost was tuned via:
- **Phase 1**: RandomizedSearchCV (50 iterations × 3-fold CV = 150 fits) — broad exploration
- **Phase 2**: GridSearchCV (30 combinations × 3-fold CV = 90 fits) — fine-tuning top 3 parameters

This two-phase approach balances exploration efficiency with optimization precision.

### 4. Interpretability via SHAP

XGBoost + SHAP TreeExplainer provides:
- **Fast exact SHAP values** (polynomial time for tree models)
- **Local explanations** — why *this specific car* was priced at *this specific amount*
- **Global explanations** — which features drive pricing overall
- **Business-friendly** narratives for stakeholders

### 5. Production-Readiness

- Fast inference (milliseconds per prediction)
- Handles missing values natively
- Efficient memory usage
- Well-supported in MLflow, Docker, and cloud deployment platforms

---
## 10 · Launching the MLflow UI

MLflow provides a **web-based UI** to visually explore experiment runs, compare metrics, view artifacts, and manage the model registry.

### How to Launch

Open a terminal/command prompt, navigate to the project directory (where the `mlflow.db` file is located), and run:

```bash
mlflow ui --backend-store-uri sqlite:///mlflow.db
```

This starts a local web server. Open your browser and navigate to:

```
http://127.0.0.1:5000
```

### What You'll See

| UI Section | What It Shows |
|------------|---------------|
| **Experiments** | List of all experiments (select `Used_Car_Price_Prediction`) |
| **Runs Table** | All runs with metrics, parameters, and tags — sortable and filterable |
| **Run Detail** | Click any run to see its parameters, metrics, and artifacts |
| **Artifacts** | Browse logged plots (feature importance, SHAP) and model files |
| **Compare** | Select multiple runs and click "Compare" for side-by-side analysis |
| **Models** | View registered models, versions, and stage transitions |

### Advanced Options

```bash
# Specify a custom host and port
mlflow ui --backend-store-uri sqlite:///mlflow.db --host 0.0.0.0 --port 8080

# Use a specific artifact root
mlflow ui --backend-store-uri sqlite:///mlflow.db --default-artifact-root ./mlflow-artifacts
```

### For Remote/Team Access

In a production or team setting, deploy MLflow as a **tracking server**:

```bash
mlflow server \
    --host 0.0.0.0 \
    --port 5000 \
    --backend-store-uri postgresql://user:password@host:5432/mlflow \
    --default-artifact-root s3://your-bucket/mlflow-artifacts
```

Then point all notebooks to this server:
```python
mlflow.set_tracking_uri('http://your-server:5000')
```

In [13]:
# ── Display the command to launch MLflow UI ──
print('=' * 65)
print('  HOW TO LAUNCH THE MLflow UI')
print('=' * 65)
print()
print('  1. Open a terminal / command prompt')
print()
print('  2. Navigate to the project directory:')
print('     cd "d:\\MLPs\\Used Car Price Prediction & Valuation System"')
print()
print('  3. Run the MLflow UI server:')
print('     mlflow ui --backend-store-uri sqlite:///mlflow.db')
print()
print('  4. Open your browser and go to:')
print('     http://127.0.0.1:5000')
print()
print('  5. Select the experiment: "Used_Car_Price_Prediction"')
print()
print('=' * 65)
print()
print('  📊 You can compare all 3 runs side-by-side,')
print('     view logged plots, and inspect model artifacts.')
print('=' * 65)

  HOW TO LAUNCH THE MLflow UI

  1. Open a terminal / command prompt

  2. Navigate to the project directory:
     cd "d:\MLPs\Used Car Price Prediction & Valuation System"

  3. Run the MLflow UI server:
     mlflow ui --backend-store-uri sqlite:///mlflow.db

  4. Open your browser and go to:
     http://127.0.0.1:5000

  5. Select the experiment: "Used_Car_Price_Prediction"


  📊 You can compare all 3 runs side-by-side,
     view logged plots, and inspect model artifacts.


---
## 11 · Final Summary

### All Tracked Experiments

In [15]:
# ── Final Summary Table ──
summary_data = []

for run_id, run_name, metrics, t_time, stage in [
    (lr_run_id,  RUN_NAME_LR,  lr_test_metrics,  lr_training_time,  'baseline'),
    (xgb_run_id, RUN_NAME_XGB, xgb_test_metrics, xgb_training_time, 'champion'),
]:
    summary_data.append({
        'Model': run_name,
        'Run ID': run_id[:8] + '...',
        'Test MAE': f'₹{metrics["MAE"]:,.2f}',
        'Test RMSE': f'₹{metrics["RMSE"]:,.2f}',
        'Test R²': f'{metrics["R2"]:.4f}',
        'Training Time': f'{t_time:.2f}s',
        'Stage': stage.upper(),
    })

summary_df = pd.DataFrame(summary_data)

print('\n' + '=' * 110)
print('  📋 FINAL EXPERIMENT SUMMARY — All MLflow-Tracked Runs')
print('=' * 110)
print(summary_df.to_string(index=False))
print('=' * 110)


  📋 FINAL EXPERIMENT SUMMARY — All MLflow-Tracked Runs
                  Model      Run ID    Test MAE   Test RMSE Test R² Training Time    Stage
      Linear Regression 45ae6670... ₹118,221.01 ₹160,143.83  0.8760         0.23s BASELINE
Tuned XGBoost Regressor 9a465c22...  ₹54,060.33  ₹83,343.83  0.9664        12.94s CHAMPION


In [16]:
# ── Champion Model Card ──
print('\n' + '=' * 70)
print('  🏆 CHAMPION MODEL — Production-Ready Summary')
print('=' * 70)
print(f'''
  Model Name          : Tuned XGBoost Regressor
  MLflow Run ID       : {xgb_run_id}
  Registry Name       : {MODEL_REGISTRY_NAME}
  Algorithm           : XGBRegressor
  Tuning Strategy     : RandomizedSearchCV → GridSearchCV

  ── Hyperparameters ──
  n_estimators        : 500
  max_depth           : 8
  learning_rate       : 0.03
  subsample           : 0.7
  colsample_bytree    : 0.9
  min_child_weight    : 1
  gamma               : 1.0
  reg_alpha            : 0.1
  reg_lambda           : 1.0

  ── Test Set Performance ──
  MAE                 : ₹{xgb_test_metrics['MAE']:>12,.2f}
  RMSE                : ₹{xgb_test_metrics['RMSE']:>12,.2f}
  R²                  :  {xgb_test_metrics['R2']:>12.4f}

  ── Training Details ──
  Dataset             : feature_engineered_car_data.csv
  Samples             : 37,750 (Train: 30,200 / Test: 7,550)
  Features            : 81
  Training Time       : {xgb_training_time:.2f} seconds

  ── MLflow Artifacts Logged ──
  ✅ Model artifact (XGBoost native format)
  ✅ Feature importance plot
  ✅ SHAP summary plot
  ✅ All hyperparameters
  ✅ Train & test metrics (MAE, RMSE, R²)
  ✅ Training time
''')
print('=' * 70)


  🏆 CHAMPION MODEL — Production-Ready Summary

  Model Name          : Tuned XGBoost Regressor
  MLflow Run ID       : 9a465c22ca5c4a80b70bd5db33f9c40e
  Registry Name       : Used_Car_Price_XGBoost_Champion
  Algorithm           : XGBRegressor
  Tuning Strategy     : RandomizedSearchCV → GridSearchCV

  ── Hyperparameters ──
  n_estimators        : 500
  max_depth           : 8
  learning_rate       : 0.03
  subsample           : 0.7
  colsample_bytree    : 0.9
  min_child_weight    : 1
  gamma               : 1.0
  reg_alpha            : 0.1
  reg_lambda           : 1.0

  ── Test Set Performance ──
  MAE                 : ₹   54,060.33
  RMSE                : ₹   83,343.83
  R²                  :        0.9664

  ── Training Details ──
  Dataset             : feature_engineered_car_data.csv
  Samples             : 37,750 (Train: 30,200 / Test: 7,550)
  Features            : 81
  Training Time       : 12.94 seconds

  ── MLflow Artifacts Logged ──
  ✅ Model artifact (XGBoost native 

---
## Conclusion

### What We Accomplished

| Task | Status |
|------|--------|
| ✅ Installed and configured MLflow with SQLite backend | Done |
| ✅ Created experiment `Used_Car_Price_Prediction` | Done |
| ✅ Trained & logged **Linear Regression** (baseline) | Done |
| ✅ Trained & logged **Random Forest Regressor** (candidate) | Done |
| ✅ Trained & logged **Tuned XGBoost Regressor** (champion) | Done |
| ✅ Tracked parameters, metrics (MAE, RMSE, R²), and training time | Done |
| ✅ Logged feature importance plots as artifacts | Done |
| ✅ Logged SHAP summary plots as artifacts | Done |
| ✅ Logged model artifacts (sklearn & XGBoost native) | Done |
| ✅ Registered champion model in MLflow Model Registry | Done |
| ✅ Created comparison tables across all runs | Done |
| ✅ Documented how to launch `mlflow ui` | Done |

### Key Takeaways

1. **MLflow transforms experimental ML into governed ML.** Every run is logged, traceable, and reproducible — critical for production systems and regulatory compliance.

2. **The Tuned XGBoost Regressor is the clear champion.** With an R² of ~0.9664, it outperforms Linear Regression and Random Forest by a significant margin on all metrics.

3. **Experiment tracking enables data-driven model selection.** Instead of guessing, we have quantitative evidence (metrics, plots, artifacts) to justify the champion model.

4. **Model versioning via the Registry ensures production safety.** Each model version is immutable, traceable, and transitions through defined lifecycle stages.

5. **This notebook is production-portable.** The same MLflow code works locally, on a team server, or on cloud platforms (Databricks, AWS SageMaker, GCP Vertex AI) — just change the tracking URI.

---

*To explore the tracked experiments visually, run `mlflow ui --backend-store-uri sqlite:///mlflow.db` from the project directory and open http://127.0.0.1:5000 in your browser.*